components/hab/src/main.rs

```diff
@@ -1651,23 +1652,23 @@ fn get_strategy_from_input(m: &ArgMatches) -> Option<UpdateStrategy> {
         .and_then(|f| UpdateStrategy::from_str(f).ok())
 }
 
-fn sup_addr_from_input(m: &ArgMatches) -> Result<SocketAddr> {
+fn sup_addr_from_input(m: &ArgMatches) -> Result<ListenCtlAddr> {
     match m.value_of("REMOTE_SUP") {
         Some(rs) => {
             let sup_addr = if rs.find(':').is_some() {
                 rs.to_string()
             } else {
-                format!("{}:{}", rs, protocol::ctl::DEFAULT_PORT)
+                format!("{}:{}", rs, ListenCtlAddr::default().port())
             };
             let addrs: Vec<SocketAddr> = match sup_addr.to_socket_addrs() {
                 Ok(addrs) => addrs.collect(),
                 Err(e) => {
                     return Err(Error::RemoteSupResolutionError(sup_addr, e));
                 }
             };
-            Ok(addrs[0])
+            Ok(ListenCtlAddr::from(addrs[0]))
```

Comment:
I realize the `addrs[0]` was already here, but as long as we're changing code in this area, I think it's worth fixing up. In general, [`Index`](http://doc.rust-lang.org/1.31.0/std/ops/trait.Index.html) operations are unsafe. In this case, it's kind of OK because the CLI validation has already run, but we can structure the code better and make this more explicit:
```rust
            match sup_addr.to_socket_addrs() {
                Ok(addrs) => match addrs.collect::<Vec<_>>().as_slice() {
                    [addr] => return Ok(ListenCtlAddr::from(*addr)),
                    _ => panic!(
                        "CLI validation failure: REMOTE_SUP should contain single SocketAddr"
                    ),
                },
                Err(e) => {
                    return Err(Error::RemoteSupResolutionError(sup_addr, e));
                }
            }
```
Also, the overall structure of the function could be improved. I'd suggest:
```rust
    m.value_of("REMOTE_SUP")
        .map_or(Ok(ListenCtlAddr::default()), |rs| {
            // parse string → ListenCtlAddr
        })
```

Reply:
![tenor-82425110](https://user-images.githubusercontent.com/207178/49876715-ed996980-fdf1-11e8-8668-35386da90ab1.gif)

Reply:
I can't pattern match on addrs here because the vector can contain more than one item and still be valid. I tweaked things a bit to implement your other suggestions though.

Reply:
Sorry @baumanj, I should have left more explanation. We only allow for a single value for `--remote-sup` but we don't require that value to be an IP and thus aren't guaranteed that `to_socket_addrs` will return only a single value. If the user passes a name then `to_socket_addrs` can resolve it to both an IPv4 and IPv6 returning a vector with 2 values.

Reply:
Also, I did add a check to make sure that the vector isn't empty. I think there is still more refactoring that could be done to this function, but that should at least make this operation safer.

Reply:
Ah, I follow you now. Thanks for the clarification.

If we were further in the future, you could use this [advanced slice pattern](https://doc.rust-lang.org/1.9.0/book/slice-patterns.html) syntax:
```rust
    match addrs_slice {
        [primary, _secondaries..] => primary,
        _ => …,
    }
```
But it's not available yet, so I'd suggest this approach:
```rust
    match listen_ctl_addr.to_socket_addrs().and_then(|mut addrs| {
        addrs.next().ok_or(err))
    }) {
        Ok(addr) => addr,
        Err(e) => …,
    }
```

I've left out the details of the error handling, but you'll probably want some sort of [`io::Error`](http://doc.rust-lang.org/1.31.0/std/io/struct.Error.html#method.new) (probably [`ErrorKind::AddrNotAvailable`](http://doc.rust-lang.org/1.31.0/std/io/enum.ErrorKind.html#variant.AddrNotAvailable)) in the case that the iterator returns no values.

Additionally, this definitely merits a comment about why `to_socket_addrs` could return multiple values and why we're only paying attention to the first one.